server/sonar-server/src/main/java/org/sonar/server/permission/ws/UsersAction.java

```diff
@@ -90,42 +87,49 @@ public void define(WebService.NewController context) {
-  }
-
-  private UsersWsResponse doHandle(UsersWsRequest request) {
-    Optional<WsProjectRef> wsProjectRef = newOptionalWsProjectRef(request.getProjectId(), request.getProjectKey());
-    validatePermission(request.getPermission(), wsProjectRef);
+  public void handle(Request request, Response response) throws Exception {
     DbSession dbSession = dbClient.openSession(false);
     try {
-      Optional<ComponentDto> project = dependenciesFinder.searchProject(dbSession, wsProjectRef);
-      checkProjectAdminUserByComponentDto(userSession, project);
-      PermissionQuery dbQuery = buildPermissionQuery(request, project);
-      List<UserDto> users = findUsers(dbSession, dbQuery);
-      int total = dbClient.userPermissionDao().countUsers(dbSession, dbQuery);
-      List<ExtendedUserPermissionDto> userPermissions = findUserPermissions(dbSession, users, project);
-      Paging paging = Paging.forPageIndex(request.getPage()).withPageSize(request.getPageSize()).andTotal(total);
-      return buildResponse(users, userPermissions, paging);
+      Optional<ProjectId> projectId = support.findProject(dbSession, request);
+      checkProjectAdminUserByComponentUuid(userSession, projectId.isPresent() ? projectId.get().getUuid() : null);
+
+      PermissionQuery query = buildPermissionQuery(request, projectId);
+      List<UserDto> users = findUsers(dbSession, query);
+      int total = dbClient.userPermissionDao().countUsers(dbSession, query);
+      List<ExtendedUserPermissionDto> userPermissions = findUserPermissions(dbSession, users, projectId);
+      Paging paging = Paging.forPageIndex(request.mandatoryParamAsInt(Param.PAGE)).withPageSize(query.getPageSize()).andTotal(total);
+      UsersWsResponse usersWsResponse = buildResponse(users, userPermissions, paging);
+      writeProtobuf(usersWsResponse, request, response);
     } finally {
       dbClient.closeSession(dbSession);
     }
   }
 
-  private static UsersWsRequest toUsersWsRequest(Request request) {
-    UsersWsRequest usersRequest = new UsersWsRequest()
-      .setPermission(request.param(PARAM_PERMISSION))
-      .setProjectId(request.param(PARAM_PROJECT_ID))
-      .setProjectKey(request.param(PARAM_PROJECT_KEY))
-      .setQuery(request.param(Param.TEXT_QUERY))
-      .setPage(request.mandatoryParamAsInt(Param.PAGE))
-      .setPageSize(request.mandatoryParamAsInt(Param.PAGE_SIZE));
-
-    String searchQuery = usersRequest.getQuery();
-    checkRequest(searchQuery == null || searchQuery.length() >= SEARCH_QUERY_MIN_LENGTH,
-      "The '%s' parameter must have at least %d characters", Param.TEXT_QUERY, SEARCH_QUERY_MIN_LENGTH);
-    return usersRequest;
+  private static PermissionQuery buildPermissionQuery(Request request, Optional<ProjectId> project) {
+    String textQuery = request.param(Param.TEXT_QUERY);
+    String permission = request.param(PARAM_PERMISSION);
+    PermissionQuery.Builder permissionQuery = PermissionQuery.builder()
```

Comment:
creating variable `permissionQuery` could/should go after checking the permission (avoid to create useless variable and execute useless code when check fails)